src/sentry/rules/processing/delayed_processing.py

```diff
@@ -219,5 +252,32 @@ def apply_delayed(project_id: int) -> DefaultDict[Rule, set[int]] | None:
         rules_to_fire = get_rules_to_fire(
             condition_group_results, rule_to_slow_conditions, rules_to_groups
         )
-        return rules_to_fire
-    return None
+    # Step 7: Ready, aim, fire!!
+    group_id_to_event = get_group_id_to_event(rulegroup_to_events, project)
+
+    now = timezone.now()
+    for rule, group_ids in rules_to_fire.items():
+        frequency = rule.data.get("frequency") or Rule.DEFAULT_FREQUENCY
+        freq_offset = now - timedelta(minutes=frequency)
+        groups = Group.objects.filter(id__in=group_ids)
+        for group in groups:
+            rule_statuses = bulk_get_rule_status(alert_rules, group, project)
+            status = rule_statuses[rule.id]
+            if status.last_active and status.last_active > freq_offset:
+                return
+
+            updated = (
+                GroupRuleStatus.objects.filter(id=status.id)
+                .exclude(last_active__gt=freq_offset)
+                .update(last_active=now)
+            )
+
+            if not updated:
+                return
+
+            event = group_id_to_event[group.id]
+            notification_uuid = str(uuid.uuid4())
+            rule_fire_history = history.record(rule, group, event.event_id, notification_uuid)
+            activate_downstream_actions(
+                rule, event.for_group(group), notification_uuid, rule_fire_history
```

Comment:
We'll also need to get the occurrence for the error here if this is an issue platform event. We should have this available in post process, so maybe the best option is to include the occurrence id along with the event id, like `<event_id>::<occurrence_id` and then we just decode it.

The other option is to query snuba to get it, but we're probably better off just passing it along.

We could probably do the `for_group` stuff in `get_group_id_to_event` potentially, so that we just have a bunch of `GroupEvent` objects with/without occurrences ready to go.

Reply:
No, only issue platform events.

Starting to wonder if we should just be storing a json blob in the value. That way we can store event_id, occurrence_id, and any other things we might need in the future.

Reply:
I like the idea of storing this in JSON if we have to add more fields if we need to add more fields.

---

Is there anywhere I can dive into to figure out all the differences between events (Error, issue platform, etc)? Just the event code itself?

Reply:
The main different is that an error event has no `occurrence` associated with it, and issue platform events do. Occurrences are basically extra metadata that gives us info about the specific event type. So for an N+1 db query perf issue it'd have info about the query that was repeated.

https://develop.sentry.dev/issue-platform/
https://www.notion.so/sentry/Issue-Platform-Overview-b5799988da494f11a435479ffb539391